### PR TITLE
Create button guide

### PR DESCRIPTION
**Pull request by gregoryb-at-ibm**
Hi

Is this appropriate for the guides repository?

It's a guide on where to add what code to enable creation of an externally managed item in your EMS.

* Need to determine location and name of file.

### CODE REVIEW

**Comment by Fryguy** [comment]
> Is this appropriate for the guides repository?

yes, this is the place!

However, I wonder if this might be better in the provider plugins directory as opposed to UI.  @agrare Thoughts on organization?

**Comment by agrare** [comment]
> However, I wonder if this might be better in the provider plugins directory

Yeah this could actually be a great next "article" in the writing a new provider series, e.g. https://github.com/ManageIQ/guides/blob/master/providers/writing_a_new_provider.md#next-steps

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:40]: This is usually handled when adding the new model, not when adding create support.  Maybe better to split up the "How to add a new model" into its own guide which would cover [[lines 22 - 40]](https://github.com/ManageIQ/guides/pull/478/files#diff-7b7fe923f33307d7a91073faa2e44c3b61e1c9c60a8e0db971e67ea7e7a960c6R22-R40)

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:48]: `options dict`

In ruby this is more commonly called a hash, python has dictionaries

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:55]: `manageiq-providers-autosde` => best to leave this generic as well like `manageiq-providers-your_provider`

Comment (agrare) [reply ui/flow_of_allowing_to_add_models.md:55]: Actually with ``supports_attribute :supports_create_your_item, :child_model => "YourItem"` the correct way to do this would be:
`manageiq-providers-your_provider/app/models/manageiq/providers/your-provider/your_manager/your_thing.rb` and add `supports :create`

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:57]: cc @kavyanekkalapu can you review the UI parts of this?

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:108]: If you're adding create to the collection you'll also need to edit `api.yml`, think it'd be `:verbs: *gpppd` but that always confuses me :laughing: assume it is `GET, PUT, PATCH, POST, DESTROY` but what if you only have `GET, POST` is it `*gp`?  How do we know that's not `GET, PATCH`? cc @Fryguy

Comment (agrare) [reply ui/flow_of_allowing_to_add_models.md:108]: NOTE you also might need to create a new product feature

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:117]: This would be a good place to mention the product_feature/identifier that you have to authorize the user for in the specs

**Comment by agrare** [comment]
I love this @gregoryb-at-ibm thanks so much for putting the whole flow together!

File: ui/flow_of_allowing_to_add_models.md
Comment (kavyanekkalapu) [reply ui/flow_of_allowing_to_add_models.md:57]: @gregoryb-at-ibm UI section Looks good to me.

File: ui/flow_of_allowing_to_add_models.md
Comment (Fryguy) [reply ui/flow_of_allowing_to_add_models.md:108]: yes, `gpppd` is `GET POST PUT PATCH DESTROY`.  The star at the beginning of `*gpppd` is a YAML anchor that points to the definition.  See https://github.com/ManageIQ/manageiq-api/blob/master/config/api.yml#L6-L31  We know it's not `GET PATCH`, because `GET PATCH` is not defined 😉

Comment (agrare) [reply ui/flow_of_allowing_to_add_models.md:108]: :+1: so if you wanted to implement only GET and PATCH you could call them out and just not use the shorthand

Comment (Fryguy) [reply ui/flow_of_allowing_to_add_models.md:108]: If it helps (or for others that aren't aware of YAML anchors)...

```ruby
[1] pry(main)> require "yaml"
=> true
[2] pry(main)> x = {:target => {:a => 1, :b => 2}}
=> {:target=>{:a=>1, :b=>2}}
[3] pry(main)> x[:reference] = x[:target]
=> {:a=>1, :b=>2}
[4] pry(main)> x[:reference].object_id == x[:target].object_id
=> true
[5] pry(main)> puts x.to_yaml
---
:target: &1
  :a: 1
  :b: 2
:reference: *1
=> nil
[6] pry(main)> y = YAML.load(x.to_yaml)
=> {:target=>{:a=>1, :b=>2}, :reference=>{:a=>1, :b=>2}}
[7] pry(main)> y[:reference].object_id == y[:target].object_id
=> true
```

Comment (chessbyte) [reply ui/flow_of_allowing_to_add_models.md:108]: so we created this nomenclature to minimize typing??

Comment (chessbyte) [reply ui/flow_of_allowing_to_add_models.md:108]: anchors are also used by Rails in `config/database.yml`, if I recall correctly

Comment (Fryguy) [reply ui/flow_of_allowing_to_add_models.md:108]: It minimizes _a lot_ of duplication, but yes, it was created to declutter the file.  It also ensures that only known "sets" are chosen.

Comment (Fryguy) [reply ui/flow_of_allowing_to_add_models.md:108]: For reference:

```ruby
[8] pry(main)> YAML.load_file("config/api.yml").deep_dup.to_yaml.lines.count
=> 5457
[9] pry(main)> File.read("config/api.yml").lines.count
=> 5068
```

The first line is fully expanded, the second line is the current, so we're talking 389 lines or 7.5% bigger

Comment (gregoryb-at-ibm) [reply ui/flow_of_allowing_to_add_models.md]: @agrare, @Fryguy 
I added instructions about adding `create` to YAML. Could you see that it's correct and complete?

Also, could you help me with instructions for adding an appropriate product feature?

File: ui/flow_of_allowing_to_add_models.md
Comment (gregoryb-at-ibm) [reply ui/flow_of_allowing_to_add_models.md:137]: I've added a section about product features. Could you check it?

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:29]: suggest changing `record` -> `base class` or `base model`

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:33]: Do you want to remove [[this]](https://github.com/ManageIQ/guides/pull/478/files#diff-7b7fe923f33307d7a91073faa2e44c3b61e1c9c60a8e0db971e67ea7e7a960c6R13) from the assumptions if you're also covering it here?  Think it should be in one place or the other but not both

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:47]: Only need to talk about adding `supports :create` once, [[ref]](https://github.com/ManageIQ/guides/pull/478/files#diff-7b7fe923f33307d7a91073faa2e44c3b61e1c9c60a8e0db971e67ea7e7a960c6R36)

File: ui/flow_of_allowing_to_add_models.md
Comment (agrare) [inline ui/flow_of_allowing_to_add_models.md:95]: Really minor, but this should be `yaml` not `shell`

**Comment by agrare** [comment]
Looking good @gregoryb-at-ibm just a few minor comments then I'll merge!

File: ui/flow_of_allowing_to_add_models.md
Comment (gregoryb-at-ibm) [reply ui/flow_of_allowing_to_add_models.md]: > Looking good @gregoryb-at-ibm just a few minor comments then I'll merge!

I think I fixed all the comments?

Before you merge we should decide on the file's name, and where it should go.
And also I'd like to squash all my commits.

**Comment by gregoryb-at-ibm** [comment]
@agrare 
I think I fixed all the comments?

Before you merge we should decide on the file's name, and where it should go.
Then I'll squash all my commits.

**Comment by agrare** [comment]
> Before you merge we should decide on the file's name, and where it should go.

How about `providers/create_operation.md`

**Comment by gregoryb-at-ibm** [comment]
> > Before you merge we should decide on the file's name, and where it should go.
> 
> How about `providers/create_operation.md`

OK, moved and squashed.

**Review by agrare — APPROVED** [review]

**Pull request merged**
